src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClients/FailOverSupportedConfigurationClient.cs

```diff
@@ -0,0 +1,253 @@
+        }
+
+        private async Task<Response<T>> ExecuteWithFailOverPolicyAsync<T>(Func<ConfigurationClient, Task<Response<T>>> funcToExecute, bool forceTryFailedReplicas = false, CancellationToken cancellationToken = default)
+        {
+            var tasks = new List<Task>();
+            var attemptsCancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            int replicasTried = 0;
+            Exception lastException = null;
+
+            for(int i = 0; i < _configurationClientAndStates.Count(); i++)
+            {
+                ConfigurationClient client = _configurationClientAndStates.ElementAt(i).Item1;
+                ConfigurationClientState clientState = _configurationClientAndStates.ElementAt(i).Item2;
+
+                if (forceTryFailedReplicas || clientState.IsAvailable())
+                {
+                    var parallelAwait = Task.Delay(this._parallelRetryInterval, attemptsCancellation.Token);
+                    Task<Response<T>> funcTask = funcToExecute(client);
+                    tasks.Add(funcTask);
+                    tasks.Add(parallelAwait);
+                    replicasTried++;
+
+                    var completedTask = await Task.WhenAny(tasks);
+                    if (!completedTask.Equals(parallelAwait))
+                    {
+                        if (completedTask.Status == TaskStatus.RanToCompletion)
+                        {
+                            clientState.UpdateConfigurationStoreStatus(requestSuccessful: true);
+
+                            // Safe because the task is completed.
+                            return funcTask.Result;
+                        }
+                        else if (completedTask.Status == TaskStatus.Faulted)
+                        {
+                            tasks.Clear();
+
+                            if (IsRetryableException(completedTask.Exception))
+                            {
+                                lastException = completedTask.Exception;
+                                clientState.UpdateConfigurationStoreStatus(requestSuccessful: false);
+                            }
+                            else
+                            {
+                                throw completedTask.Exception;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        tasks.Remove(parallelAwait);
+                    }
+                }
+            }
+
+            if (replicasTried == 0)
+            {
+                return await ExecuteWithFailOverPolicyAsync(funcToExecute, forceTryFailedReplicas: true, cancellationToken);
+            }
+
+            throw lastException;
+        }
+
+        private async Task<T> ExecuteWithFailOverPolicyAsync<T>(Func<ConfigurationClient, T> funcToExecute, bool forceTryFailedReplicas = false, CancellationToken cancellationToken = default)
```

Comment:
This overload doesn't make sense to me. The function being passed in doesn't return a task.

Reply:
Yes, the function passed doesn't return a task, but we need to asynchronously wait and initiate a parallel request.
We can't asynchronously wait unless we make this function an async function which returns a task.

I am open to suggestions to simplify it.

Reply:
There's no point in asynchronously waiting a synchronous function.

Reply:
If a synchronous function is taking very long time, shouldn't we try the same function with another configuration client?
If we want to do that, I think we would have to wrap the synchronous function inside an async function.

Let's hop over a call to discuss in detail.